File: src/main/java/me/tomasan7/jecnadesktop/JecnaDesktop.java
package me.tomasan7.jecnadesktop;

import javafx.application.Application;
import javafx.application.Platform;
import javafx.scene.image.Image;
import javafx.stage.Stage;
import me.tomasan7.jecnadesktop.repository.*;
import me.tomasan7.jecnadesktop.ui.JDScene;
import me.tomasan7.jecnadesktop.ui.SceneManager;
import me.tomasan7.jecnadesktop.ui.scene.LoadingScene;
import me.tomasan7.jecnadesktop.ui.scene.LoginScene;
import me.tomasan7.jecnadesktop.ui.scene.MainScene;
import me.tomasan7.jecnadesktop.web.JecnaWebClient;

/**
 * Main class of the application.
 */
public class JecnaDesktop extends Application
{
	private Stage primaryStage = null;
	private JecnaWebClient jecnaWebClient = null;
	private GradesRepository gradesRepository = null;
	private AttendancesRepository attendancesRepository = null;
	private TimetableRepository timetableRepository = null;
	private SceneManager sceneManager = null;

	@Override
	public void start (Stage primaryStage)
	{
		this.primaryStage = primaryStage;

		primaryStage.setTitle("Ječná Desktop");
		primaryStage.getIcons().add(new Image("/ui/icon/logo.png"));
		/* For some reason the process doesn't exit by itself, so I kill it when the stage is closed. */
		primaryStage.setOnCloseRequest(__ -> {Platform.exit(); System.exit(0);});

		sceneManager = new SceneManager(primaryStage);
		registerPages();
		sceneManager.switchToScene(JDScene.LOGIN);

		primaryStage.show();
	}

	private void registerPages ()
	{
		sceneManager.addScene(JDScene.LOGIN, new LoginScene(this));
		sceneManager.addScene(JDScene.LOADING, new LoadingScene(this));
		sceneManager.addScene(JDScene.MAIN, new MainScene(this));
	}

	public Stage getPrimaryStage ()
	{
		return primaryStage;
	}

	public JecnaWebClient getJecnaWebClient ()
	{
		return jecnaWebClient;
	}

	public void initDataAccess (JecnaWebClient jecnaWebClient)
	{
		if (this.jecnaWebClient != null)
			throw new IllegalStateException("Data access has already been initialized.");

		this.jecnaWebClient = jecnaWebClient;
		this.gradesRepository = new WebGradesRepository(jecnaWebClient);
		this.attendancesRepository = new WebAttendancesRepository(jecnaWebClient);
		this.timetableRepository = new WebTimetableRepository(jecnaWebClient);
	}

	public GradesRepository getGradesRepository ()
	{
		return gradesRepository;
	}

	public AttendancesRepository getAttendancesRepository ()
	{
		return attendancesRepository;
	}

	public TimetableRepository getTimetableRepository ()
	{
		return timetableRepository;
	}

	public SceneManager getSceneManager ()
	{
		return sceneManager;
	}
}

File: src/main/kotlin/me/tomasan7/jecnadesktop/JecnaDesktop.kt
package me.tomasan7.jecnadesktop

import javafx.application.Platform
import javafx.event.EventHandler
import javafx.scene.image.Image
import javafx.stage.Stage
import me.tomasan7.jecnaapi.repository.*
import me.tomasan7.jecnaapi.web.JecnaWebClient
import me.tomasan7.jecnadesktop.ui.JDScene
import me.tomasan7.jecnadesktop.ui.SceneManager
import me.tomasan7.jecnadesktop.ui.scene.LoadingScene
import me.tomasan7.jecnadesktop.ui.scene.LoginScene
import me.tomasan7.jecnadesktop.ui.scene.MainScene
import tornadofx.App
import kotlin.system.exitProcess

/**
 * Main class of the application.
 */
class JecnaDesktop : App()
{
    var primaryStage: Stage? = null
        private set

    var jecnaWebClient: JecnaWebClient? = null
        private set

    var gradesRepository: GradesRepository? = null
        private set

    var attendancesRepository: AttendancesRepository? = null
        private set

    var timetableRepository: TimetableRepository? = null
        private set

    var sceneManager: SceneManager? = null
        private set

    override fun start(stage: Stage)
    {
        this.primaryStage = stage

        stage.title = "Ječná Desktop"
        stage.icons.add(Image("/ui/icon/logo.png"))
        /* For some reason the process doesn't exit by itself, so I kill it when the stage is closed. */
        stage.onCloseRequest = EventHandler {
            Platform.exit()
            exitProcess(0)
        }

        sceneManager = SceneManager(stage)
        registerPages()
        sceneManager!!.switchToScene(JDScene.LOGIN)

        stage.show()
    }

    private fun registerPages()
    {
        sceneManager!!.addScene(JDScene.LOGIN, LoginScene(this))
        sceneManager!!.addScene(JDScene.LOADING, LoadingScene())
        sceneManager!!.addScene(JDScene.MAIN, MainScene(this))
    }

    fun initDataAccess(jecnaWebClient: JecnaWebClient?)
    {
        checkNotNull(jecnaWebClient) { "Data access has already been initialized." }

        this.jecnaWebClient = jecnaWebClient
        gradesRepository = WebGradesRepository(jecnaWebClient)
        attendancesRepository = WebAttendancesRepository(jecnaWebClient)
        timetableRepository = WebTimetableRepository(jecnaWebClient)
    }
}